File: leetcode6.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        if(numRows <= 1 || numRows >= s.length())
          return s;
        string res;
        for (int i = 1; i <= numRows ; ++i)
        {
          int p = i - 1;
          if(i == 1 || i == numRows) //first row and the last row
          {
            while(p < s.length())
            {
              res.push_back(s[p]);
              p = p + 2 * (numRows - 1);
            }
          }
          else //other rows
          {
            int delta = 2 * (numRows - i);
            while(p < s.length())
            {
              res.push_back(s[p]);
              p = p + delta;
              delta = 2 * (numRows - 1) - delta;
            }
          }
        }
        return res;
    }
};

File: leetcode6.java
public class Solution {
    public String convert(String s, int numRows) {
        char c[] = s.toCharArray();
        int len = c.length;
        StringBuffer[] sb = new StringBuffer[numRows];
        for (int i = 0; i < numRows ;++i) {
        	sb[i] = new StringBuffer();
        }

        int i = 0;
        while(i < len){
        	for(int j = 0; j < numRows && i < len; ++j){
        		sb[j].append(c[i]);
        		i++;
        	}
        	for(int j = numRows-2; j >= 1 && i < len; --j){
        		sb[j].append(c[i]);
        		i++;
        	}
        }
        String res = "";
        for (int m = 0;m < sb.length ;m++ ) {
        	res += sb[m];
        }
        return res;
    }
}
